sdks/python/container/boot.go

```diff
@@ -200,24 +211,112 @@ func main() {
 		}
 	}
 
+	workerIds := append([]string{*id}, info.GetSiblingWorkerIds()...)
+
+	// Keep track of child PIDs for clean shutdown without zombies
+	childPids := struct {
+		v []int
+		canceled bool
+		mu sync.Mutex
+	} {v: make([]int, 0, len(workerIds))}
+
+	// Forward trapped signals to child process groups in order to terminate them gracefully and avoid zombies
+	go func() {
+		log.Printf("Received signal: %v", <-signalChannel)
+		childPids.mu.Lock()
+		childPids.canceled = true
+		for _, pid := range childPids.v {
+			go func(pid int) {
+				// This goroutine will be canceled if the main process exits before the 5 seconds
+				// have elapsed, i.e., as soon as all subprocesses have returned from Wait().
+				time.Sleep(5 * time.Second)
+				if err := syscall.Kill(-pid, syscall.SIGKILL); err == nil {
+					log.Printf("Worker process %v did not respond, killed it.", pid)
+				}
+			}(pid)
+			syscall.Kill(-pid, syscall.SIGTERM)
```

Comment:
Q: Why is the negative pid being used here? I'd have expected simply `pid`.

I haven't used syscall.Kill before, and it doesn't have any doc comment, https://pkg.go.dev/syscall#Kill, so it's very odd to me. If there's some magic here, please add a comment explaining the oddity.

Reply:
The native PID refers to the process group, so we can SIGTERM the whole tree in case the Python process spawns child processes. The process group is created further down in the code. If we don't do this, the sub processes will be orphaned, creating zombie processes if the Python process fails to wait() them.